File: aoc24/2/main.clj
(require '[clojure.string :as str])

(defn ordered? [coll]
  (or (= coll (sort coll))
      (= coll (sort > coll))))

(defn adjacent-diff? [nums]
  (every? #(and (>= % 1) (<= % 3)) (map #(abs (- %1 %2)) (rest nums) nums)))

(defn is-safe? [nums]
  (and (ordered? nums) (adjacent-diff? nums)))

(def ans (->> (slurp "in.txt")
              str/split-lines
              (map (fn [line] (map #(Integer/parseInt %) (str/split line #" "))))
              (map is-safe?)
              (filter true?)
              count))

(println ans)

(defn remove-each [nums]
  (map-indexed
   (fn [idx _]
     (concat (take idx nums) (drop (inc idx) nums)))
   nums))

(defn is-safe2? [nums]
  (some #(and (ordered? %) (adjacent-diff? %)) (remove-each nums)))

(def ans2 (->> (slurp "in.txt")
               str/split-lines
               (map (fn [line] (map #(Integer/parseInt %) (str/split line #" "))))
               (map is-safe2?)
               (filter true?)
               count))

(println ans2)

File: aoc24/2/main.py
import sys

data = open(sys.argv[1]).read().strip()
lines = data.split('\n')

ans = 0
for line in lines:
    nums = [int(n) for n in line.split()]
    if nums == sorted(nums) or nums == sorted(nums, reverse=True):
        broke = False
        for i in range(1, len(nums)):
            if not (3 >= abs(nums[i] - nums[i-1]) >= 1):
                broke = True
                break
        if not broke:
            ans += 1

print(ans)


ans = 0
for line in lines:
    nums = [int(n) for n in line.split()]
    can_be_made_safe = False
    for i, num in enumerate(nums):
        new_nums = nums[:i] + nums[i+1:]
        if new_nums == sorted(new_nums) or new_nums == sorted(new_nums, reverse=True):
            broke = False
            for j in range(1, len(new_nums)):
                if not (3 >= abs(new_nums[j] - new_nums[j-1]) >= 1):
                    broke = True
                    break
            if not broke:
                can_be_made_safe = True
                break
    if can_be_made_safe:
        ans += 1
print(ans)
